ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java

```diff
@@ -241,9 +241,9 @@ private MemoizedSupplier<List<CompletableFuture<Message>>> applyLog() throws Raf
         final CompletableFuture<Message> f = server.applyLogToStateMachine(next);
         final long incremented = appliedIndex.incrementAndGet(debugIndexChange);
         Preconditions.assertTrue(incremented == nextIndex);
+        notifyAppliedIndex(incremented);
         if (f != null) {
           futures.get().add(f);
-          f.thenAccept(m -> notifyAppliedIndex(incremented));
```

Comment:
Let's keep this line and add `notifyAppliedIndex(incremented)` in `else`?

Reply:
Thanks for the advice! We should wait until the future completes.